exonum/src/runtime/execution_context.rs

```diff
@@ -0,0 +1,416 @@
+    /// Hash of the currently executing transaction, or `None` for non-transaction calls.
+    transaction_hash: Option<Hash>,
+    /// Reference to the dispatcher.
+    dispatcher: &'a Dispatcher,
+    /// Depth of the call stack.
+    call_stack_depth: u64,
+}
+
+impl<'a> ExecutionContext<'a> {
+    /// Maximum depth of the call stack.
+    pub const MAX_CALL_STACK_DEPTH: u64 = 128;
+
+    pub(crate) fn for_transaction(
+        dispatcher: &'a Dispatcher,
+        fork: &'a mut Fork,
+        instance: InstanceDescriptor<'a>,
+        author: PublicKey,
+        transaction_hash: Hash,
+    ) -> Self {
+        Self::new(
+            dispatcher,
+            fork,
+            instance,
+            Caller::Transaction { author },
+            Some(transaction_hash),
+        )
+    }
+
+    pub(crate) fn for_block_call(
+        dispatcher: &'a Dispatcher,
+        fork: &'a mut Fork,
+        instance: InstanceDescriptor<'a>,
+    ) -> Self {
+        Self::new(dispatcher, fork, instance, Caller::Blockchain, None)
+    }
+
+    fn new(
+        dispatcher: &'a Dispatcher,
+        fork: &'a mut Fork,
+        instance: InstanceDescriptor<'a>,
+        caller: Caller,
+        transaction_hash: Option<Hash>,
+    ) -> Self {
+        Self {
+            dispatcher,
+            fork,
+            instance,
+            caller,
+            transaction_hash,
+            interface_name: "",
+            call_stack_depth: 0,
+        }
+    }
+
+    /// Returns the hash of the currently executing transaction, or `None` for non-transaction
+    /// root calls (e.g., `before_transactions` / `after_transactions` service hooks).
+    pub fn transaction_hash(&self) -> Option<Hash> {
+        self.transaction_hash
+    }
+
+    /// Provides access to blockchain data.
+    pub fn data(&self) -> BlockchainData<'a, &Fork> {
+        BlockchainData::new(self.fork, self.instance)
+    }
+
+    /// Provides access to the data of the executing service.
+    pub fn service_data(&self) -> Prefixed<'a, &Fork> {
+        self.data().for_executing_service()
+    }
+
+    /// Returns the authorization information about this call.
+    pub fn caller(&self) -> &Caller {
+        &self.caller
+    }
+
+    /// Returns a descriptor of the executing service instance.
+    pub fn instance(&self) -> InstanceDescriptor<'_> {
+        self.instance
+    }
+
+    /// Returns `true` if currently processed block is a genesis block.
+    pub fn in_genesis_block(&self) -> bool {
+        let core_schema = self.data().for_core();
+        core_schema.next_height() == Height(0)
+    }
+
+    /// Returns identifier of the service interface required for the call.
+    ///
+    /// Keep in mind that this getter, in fact, is a part of an unfinished "interfaces" feature.
+    /// It will be replaced in future releases. At the moment this field is always empty for
+    /// the primary service interface.
+    pub fn interface_name(&self) -> &str {
```

Comment:
Nit: shouldn't it be `doc(hidden)` then?

Reply:
Interface name is used in the sample runtime example. So it is a little bit strange to use hidden features in the sample example.